src/test/dup/test_dup.rs

```diff
@@ -0,0 +1,85 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+use test_utils::set;
+use test_utils::TestEnvironment as TestEnv;
+
+fn main() -> Result<(), String> {
+    // should we restrict the tests we run?
+    let filter_shadow_passing = std::env::args().any(|x| x == "--shadow-passing");
+    let filter_libc_passing = std::env::args().any(|x| x == "--libc-passing");
+    // should we summarize the results rather than exit on a failed test
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let mut tests = get_tests();
+    if filter_shadow_passing {
+        tests = tests
+            .into_iter()
+            .filter(|x| x.passing(TestEnv::Shadow))
+            .collect()
+    }
+    if filter_libc_passing {
+        tests = tests
+            .into_iter()
+            .filter(|x| x.passing(TestEnv::Libc))
+            .collect()
+    }
+
+    test_utils::run_tests(&tests, summarize)?;
+
+    println!("Success.");
+    Ok(())
+}
+
+fn get_tests() -> Vec<test_utils::ShadowTest<(), String>> {
+    let tests: Vec<test_utils::ShadowTest<_, _>> = vec![
+        test_utils::ShadowTest::new("test_dup", test_dup, set![TestEnv::Libc, TestEnv::Shadow]),
+        test_utils::ShadowTest::new("test_dup2", test_dup2, set![TestEnv::Libc, TestEnv::Shadow]),
+        test_utils::ShadowTest::new("test_dup3", test_dup3, set![TestEnv::Libc, TestEnv::Shadow]),
+    ];
+
+    tests
+}
+
+fn test_dup() -> Result<(), String> {
+    let (read_fd, write_fd) = nix::unistd::pipe().unwrap();
+
+    test_utils::run_and_close_fds(&[write_fd, read_fd], || {
+        let write_fd_dup =
+            test_utils::check_system_call!(|| { unsafe { libc::dup(write_fd) } }, &[])?;
```

Comment:
Seems worth validating the end state a bit more. e.g. write to both the old and new descriptor and validate that data from both shows up on the read end? Same for other tests below.

Reply:
The pipe tests already had this same test, but I moved it to the dup tests and run it for all dup(), dup2(), and dup3().